### ISSUE TITLE
module unload does not undo all changes

### ISSUE THREAD

**Issue by UweSauter** [TACC/Lmod#566]
```
14:29:28 hpcoft30:hpc43203@cl1fr1 ~ 
$ env > env.start.txt

14:29:32 hpcoft30:hpc43203@cl1fr1 ~ 
$ module --version

Modules based on Lua: Version 8.6.11  2022-02-03 16:59 -04:00
    by Robert McLay mclay@tacc.utexas.edu


14:29:36 hpcoft30:hpc43203@cl1fr1 ~ 
$ module load compiler/intel

14:29:46 hpcoft30:hpc43203@cl1fr1 ~ 
$ env > env.intel_loaded.txt

14:29:54 hpcoft30:hpc43203@cl1fr1 ~ 
$ module unload compiler/intel

14:30:03 hpcoft30:hpc43203@cl1fr1 ~ 
$ env > env.intel_unloaded.txt

14:31:52 hpcoft30:hpc43203@cl1fr1 ~ 
$ diff -U 1 env.start.txt env.intel_unloaded.txt
--- env.start.txt		2022-02-14 14:29:32.183596000 +0100
+++ env.intel_unloaded.txt	2022-02-14 14:30:08.114450000 +0100
@@ -2,3 +2,3 @@
 LS_COLORS=rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:mi=01;05;37;41:su=37;41:sg=30;43:ca=30;41:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.tar=01;31:*.tgz=01;31:*.arc=01;31:*.arj=01;31:*.taz=01;31:*.lha=01;31:*.lz4=01;31:*.lzh=01;31:*.lzma=01;31:*.tlz=01;31:*.txz=01;31:*.tzo=01;31:*.t7z=01;31:*.zip=01;31:*.z=01;31:*.dz=01;31:*.gz=01;31:*.lrz=01;31:*.lz=01;31:*.lzo=01;31:*.xz=01;31:*.zst=01;31:*.tzst=01;31:*.bz2=01;31:*.bz=01;31:*.tbz=01;31:*.tbz2=01;31:*.tz=01;31:*.deb=01;31:*.rpm=01;31:*.jar=01;31:*.war=01;31:*.ear=01;31:*.sar=01;31:*.rar=01;31:*.alz=01;31:*.ace=01;31:*.zoo=01;31:*.cpio=01;31:*.7z=01;31:*.rz=01;31:*.cab=01;31:*.wim=01;31:*.swm=01;31:*.dwm=01;31:*.esd=01;31:*.jpg=01;35:*.jpeg=01;35:*.mjpg=01;35:*.mjpeg=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.png=01;35:*.svg=01;35:*.svgz=01;35:*.mng=01;35:*.pcx=01;35:*.mov=01;35:*.mpg=01;35:*.mpeg=01;35:*.m2v=01;35:*.mkv=01;35:*.webm=01;35:*.ogm=01;35:*.mp4=01;35:*.m4v=01;35:*.mp4v=01;35:*.vob=01;35:*.qt=01;35:*.nuv=01;35:*.wmv=01;35:*.asf=01;35:*.rm=01;35:*.rmvb=01;35:*.flc=01;35:*.avi=01;35:*.fli=01;35:*.flv=01;35:*.gl=01;35:*.dl=01;35:*.xcf=01;35:*.xwd=01;35:*.yuv=01;35:*.cgm=01;35:*.emf=01;35:*.ogv=01;35:*.ogx=01;35:*.aac=01;36:*.au=01;36:*.flac=01;36:*.m4a=01;36:*.mid=01;36:*.midi=01;36:*.mka=01;36:*.mp3=01;36:*.mpc=01;36:*.ogg=01;36:*.ra=01;36:*.wav=01;36:*.oga=01;36:*.opus=01;36:*.spx=01;36:*.xspf=01;36:
-__LMOD_REF_COUNT_PATH=/opt/system/wrappers:1;/opt/system/ws/1.3.7/bin:1;/usr/local/bin:1;/usr/bin:1;/usr/local/sbin:1;/usr/sbin:1;/opt/system/pbs/20201009/bin:1
+__LMOD_REF_COUNT_PATH=/zhome/academic/HLRS/hlrs/hpcoft30/.local/bin:1;/zhome/academic/HLRS/hlrs/hpcoft30/bin:2;/opt/system/wrappers:1;/opt/system/ws/1.3.7/bin:1;/usr/local/bin:1;/usr/bin:1;/usr/local/sbin:1;/usr/sbin:1;/opt/system/pbs/20201009/bin:1
 _ModuleTable002_=c3RlbS9tb2R1bGVmaWxlcy9zeXN0ZW0vYmF0Y2hzeXN0ZW0vYXV0byIsCmZ1bGxOYW1lID0gInN5c3RlbS9iYXRjaHN5c3RlbS9hdXRvIiwKbG9hZE9yZGVyID0gMywKcHJvcFQgPSB7fSwKc3RhY2tEZXB0aCA9IDAsCnN0YXR1cyA9ICJhY3RpdmUiLAp1c2VyTmFtZSA9ICJzeXN0ZW0vYmF0Y2hzeXN0ZW0iLAp3ViA9ICIqYXV0by4qemZpbmFsIiwKfSwKWyJzeXN0ZW0vcGJzIl0gPSB7CmZuID0gIi9vcHQvc3lzdGVtL21vZHVsZWZpbGVzL3N5c3RlbS9wYnMvMjAyMDEwMDkiLApmdWxsTmFtZSA9ICJzeXN0ZW0vcGJzLzIwMjAxMDA5IiwKbG9hZE9yZGVyID0gMiwKcHJvcFQgPSB7fSwKc3RhY2tEZXB0aCA9IDEsCnN0YXR1cyA9ICJhY3RpdmUiLAp1c2VyTmFtZSA9ICJzeXN0
@@ -34,2 +34,3 @@
 LOADEDMODULES=hlrs-stack/2022-Q1:system/pbs/20201009:system/batchsystem/auto:system/site_names:system/ws/1.3.7:system/wrappers/1.0
+_ModuleTable006_=eXN0ZW1CYXNlTVBBVEggPSAiL29wdC9zeXN0ZW0vbW9kdWxlZmlsZXM6L29wdC9obHJzL21vZHVsZWZpbGVzIiwKfQo=
 __LMOD_REF_COUNT_MANPATH=/opt/system/ws/1.3.7/man:1;/opt/system/lmod/8.6.11/share/man:1;/opt/system/pbs/20201009/share/man:1
@@ -51,3 +52,3 @@
 MODULEPATH_ROOT=/opt/system/modulefiles
-PATH=/zhome/academic/HLRS/hlrs/hpcoft30/.local/bin:/zhome/academic/HLRS/hlrs/hpcoft30/bin:/zhome/academic/HLRS/hlrs/hpcoft30/bin:/opt/system/wrappers:/opt/system/ws/1.3.7/bin:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/opt/system/pbs/20201009/bin
+PATH=/zhome/academic/HLRS/hlrs/hpcoft30/.local/bin:/zhome/academic/HLRS/hlrs/hpcoft30/bin:/opt/system/wrappers:/opt/system/ws/1.3.7/bin:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/opt/system/pbs/20201009/bin
 _LMFILES_=/opt/hlrs/modulefiles/hlrs-stack/2022-Q1.lua:/opt/system/modulefiles/system/pbs/20201009:/opt/system/modulefiles/system/batchsystem/auto:/opt/system/modulefiles/system/site_names:/opt/system/modulefiles/system/ws/1.3.7:/opt/system/modulefiles/system/wrappers/1.0
```

It seems that not all changes to the environment are undone when a module is unloaded. I can see this with different module files. They have in common that they are still TCL files, e.g.
```
$ cat intel/19.1.3 
#%Module1.0
#
# intel compiler module 19.1.3
#

conflict compiler

set BASE_PREFIX /opt/hlrs/non-spack
set RELEASE 2022-02   
set CAT compiler
set PACKAGE intel
set COMP_VERSION 19.1.3
set VERSION 19.1.3.304/compilers_and_libraries_2020.4.304/linux
set GCC_VERSION "11.2.0"

set INTEL_BASE $BASE_PREFIX/$RELEASE/$CAT/$PACKAGE/$VERSION

setenv SITE_INTEL_HOME ${INTEL_BASE}
setenv INTEL_VERSION   ${COMP_VERSION}

# use GCC libs and includes 
prepend-path   PATH             $BASE_PREFIX/$RELEASE/$CAT/gcc/$GCC_VERSION/bin
prepend-path   LD_LIBRARY_PATH  $BASE_PREFIX/$RELEASE/$CAT/gcc/$GCC_VERSION/lib64/
prepend-path   LD_LIBRARY_PATH  $BASE_PREFIX/$RELEASE/$CAT/gcc/$GCC_VERSION/lib/

prepend-path  PATH               $INTEL_BASE/bin/intel64
prepend-path  LD_LIBRARY_PATH    $INTEL_BASE/compiler/lib/intel64
prepend-path  MANPATH            $INTEL_BASE/man/common:$BASE_PREFIX/$RELEASE/$CAT/gcc/$GCC_VERSION/share/man
prepend-path  LM_LICENSE_FILE    "26000@X.X.X.X"
# -----------------------------------------------------------------
## intel expects error messages on english
setenv        LC_MESSAGES      en_US.UTF-8
setenv        LC_COLLATE       en_US.UTF-8
setenv        LC_CTYPE         en_US.UTF-8
# --------------------------------------------------------------------
module-whatis "Intel Parallel Studio XE for Linux version : $COMP_VERSION"
# --------------------------------------------------------------------

proc ModulesHelp { } {
        global COMP_VERSION

        puts stderr "\ticc   - invokes the Intel(R) C compiler"
        puts stderr "\ticpc  - invokes the Intel(R) C++ compiler"
        puts stderr "\tifort - invokes the Intel(R) Fortran Compiler"
        puts stderr "\n\tsee 'man icc' or 'man icpc' or 'man ifort'"
        puts stderr "\n\tVersion $COMP_VERSION\n"
}
```
[env.start.txt](https://github.com/TACC/Lmod/files/8061131/env.start.txt)
[env.intel_loaded.txt](https://github.com/TACC/Lmod/files/8061149/env.intel_loaded.txt)
[env.intel_unloaded.txt](https://github.com/TACC/Lmod/files/8061150/env.intel_unloaded.txt)

Can this behavior be attributed to the TCL files or is something wrong with Lmod?

**Comment by rtmclay** [comment]
There was an issue that Lmod would not always handle LOADEDMODULES and _LMFILES_.  This is now fixed.  This is has nothing to do with TCL modulefiles. There will be difference between the start env and end env that have to do with internal Lmod environment variables.  So changes in the module table variables are to be expected.  (_ModuleTable*_) variables.  Also variables that do path reference counting (__LMOD_REF_COUNT_*) must report the difference between the current and previous module command.

Also your intel compiler module do setenv LC_COLLATE.  This means that if the user had a value for LC_COLLATE it would be lost when the intel module is unloaded. If you want different behavior you might want to use pushenv instead of setenv for LC_COLLATE.  I as a user on your system would be unhappy that you changed my value for LC_COLLATE and wiped out my value for LC_COLLATE when unloading the intel compiler modules.

My testing uses the following setup:
```
module purge
clearMT
env | sort > start.env.txt
module load compiler/intel
module unload compiler/intel
env | sort > end.env.txt
```

Please use this setup if you find any other differences.

Please test Lmod 8.6.13 to see if works for you.

**Comment by UweSauter** [comment]
> There will be difference between the start env and end env that have to do with internal Lmod environment variables. So changes in the module table variables are to be expected. (_ModuleTable*_) variables. Also variables that do path reference counting (__LMOD_REF_COUNT_*) must report the difference between the current and previous module command.

I was aware of that. But PATH also had left-overs that should have been removed.

> Also your intel compiler module do setenv LC_COLLATE. This means that if the user had a value for LC_COLLATE it would be lost when the intel module is unloaded. If you want different behavior you might want to use pushenv instead of setenv for LC_COLLATE. I as a user on your system would be unhappy that you changed my value for LC_COLLATE and wiped out my value for LC_COLLATE when unloading the intel compiler modules.

I pointed the colleague who maintains the compiler setup at our site to this ticket and you comments.

> Please test Lmod 8.6.13 to see if works for you.

Some changes  between 8.6.11 and 8.6.13 introduced a new issue right when logging in as well as whenever I try to load a module. Will need to test if this due to configuration.

This also prevents testing if *this* issue is fixed.

**Comment by UweSauter** [comment]
I see that the issue is already reported as #571.
Looks like this ticket was fixed by 8.6.13.
